### DISCUSSION TITLE
glyphhanger - Has anyone succeeded using with KaTeX?

### DISCUSSION THREAD

**Discussion by universemaster** [KaTeX/KaTeX#3361] [General]
Hi,

Maybe someone else has had more success than me.

Despite having 16,000 katex "equations" on a page and 39 images, the fonts are still 17% of the transfer over the network (from chrome lighthouse).

So, I tried using [glyphhanger](https://github.com/zachleat/glyphhanger) and almost succeeded.

Once subset to only the glyphs I used on the page (apparently), I could reduce the 20 woff2 file sizes from 260KB to 190KB (a modest, but real, improvement).  And in retrospect, writing it here, a suspiciously small reduction.

Aside: (Also, chrome lighthouse reports 8 files and 116.0KB transfer for the unoptimised set - whereas glyphhanger gives me 20 files with a total 190KB.)

However, after my initial excitement of some success I realised that characters like ℂ, and ℤ didn't appear correctly.

I'd sunk hours into using glyphhanger, so at this point I reverted to the full set and gave up.

Has anyone had more success?

I wonder if the disappointing file size reduction was because you have to use the ttf as the base, rather than starting with the woff2?

I didn't use the unicode-ranges output by glyphanger, because they  themselves were very long.  For example, here is KaTeX_SansSerif-Regular.css:

```
unicode-range: U+20-29,U+2B-3F,U+41-5F,U+61-7B,U+7D,U+7E,U+A0,U+A8,U+AC,U+B0,U+B1,U+D7,U+F4,U+F6,U+151,U+2C6,U+2C9,U+2D9,U+394,U+3A0,U+3A3,U+3B1-3B5,U+3B7-3BE,U+3C0,U+3C1,U+3C3,U+3C4,U+3C6,U+3C7,U+3C9,U+3D5,U+3F5,U+2006,U+200B,U+2019,U+2020,U+2022,U+2026,U+2032,U+2082,U+2113,U+2124,U+2192,U+21A6,U+21D2,U+21D4,U+2200,U+2202,U+2203,U+2205,U+2207,U+2208,U+220F,U+2211-2213,U+2216-2218,U+221E,U+2220,U+2223,U+2224,U+2227-222B,U+223C,U+2245,U+2248,U+2261,U+2264,U+2265,U+226A,U+2282,U+2286,U+22B2,U+22C2-22C5,U+22EE,U+22EF,U+22F1,U+230A,U+230B,U+239B,U+239D,U+239E,U+23A0,U+23A1,U+23A3,U+23A4,U+23A6-23A9,U+23AB-23AD,U+25A1,U+25B3,U+25C1,U+25C3,U+25CB,U+2660-2663,U+27E8,U+27E9,U+27F8-27FA,U+E020,U+1D41C;
}
```

---

A final point while we are talking about fonts, when KaTeX stops supporting IE11, would it be a good idea to remove woff and ttf from katex.css because woff2 support is [so good](https://caniuse.com/?search=woff2).

**Comment by edemaine** [comment]
I guess the main thing to check is whether glyphhander is doing the correct subsetting. You can open the subsetted ttf files in https://opentype.js.org/glyph-inspector.html and check which glyphs are still present. If you see any that aren't on your site, then there's an issue (presumably with glyphhanger, not KaTeX?).

You could also try it on a much smaller example, which only uses a few characters, to see whether you get a drastic size reduction in that case.

This seems like an interesting approach to add to the documentation. Perhaps we need a new section on Optimization to collect all your ideas and approaches once they settle.
